pallets/parachain-staking/src/lib.rs

```diff
@@ -837,10 +861,152 @@ pub mod pallet {
+			} else {
+				// This message would reveal a new state inconsistency, not expected
+				log::warn!(
+					"There are more accounts in CollatorState.nominators than 
+					CollatorState.top_nominators + CollatorState.bottom_nominators for
+					account {:?}",
+					account
+				);
+			}
+		}
+		// 2. for nominator state, check if there are nominations that were inadvertently bumped
+		// -> this allows us to recover the due unreserved balances for cases of (1) that
+		// did not have the nominator state removed (it does not account for when it was removed)
+		for (account, mut state) in <NominatorState2<T>>::iter() {
+			reads += 1u64;
+			for Bond { owner, amount } in state.nominations.0.clone() {
+				reads += 1u64;
+				if <AccountsDueUnreservedBalance<T>>::get(&owner, &account).is_some() {
+					<AccountsDueUnreservedBalance<T>>::insert(&owner, &account, amount);
+					writes += 1u64;
+					if state.rm_nomination(owner).is_some() {
+						if state.nominations.0.is_empty() {
+							<NominatorState2<T>>::remove(&account);
+						} else {
+							<NominatorState2<T>>::insert(&account, state.clone());
+						}
+						writes += 1u64;
+					}
+				}
+			}
+		}
+		(reads, writes)
+	}
+
+	pub(crate) fn correct_max_nominations_per_collator_upgrade_mistake<T: Config>() -> (u64, u64) {
+		let (mut reads, mut writes) = (0u64, 0u64);
+		for (account, state) in <CollatorState2<T>>::iter() {
+			reads += 1u64;
+			// 1. collect all nominator amounts into single vec and order them
+			let mut all_nominators = state.top_nominators.clone();
+			let mut starting_bottom_nominators = state.bottom_nominators.clone();
+			all_nominators.append(&mut starting_bottom_nominators);
+			// sort all nominators from greatest to least
+			all_nominators.sort_unstable_by(|a, b| b.amount.cmp(&a.amount));
+			let top_n = T::MaxNominatorsPerCollator::get() as usize;
+			// 2. split them into top and bottom using the T::MaxNominatorsPerCollator
+			let top_nominators: Vec<Bond<T::AccountId, BalanceOf<T>>> =
+				all_nominators.clone().into_iter().take(top_n).collect();
+			let bottom_nominators = if all_nominators.len() > top_n {
+				let rest = all_nominators.len() - top_n;
+				let bottom: Vec<Bond<T::AccountId, BalanceOf<T>>> = all_nominators
+					.clone()
+					.into_iter()
+					.rev()
+					.take(rest)
+					.collect();
+				bottom
+			} else {
+				// empty, all nominations are in top
+				Vec::new()
+			};
+			let (mut total_counted, mut total_backing): (BalanceOf<T>, BalanceOf<T>) =
+				(0u32.into(), 0u32.into());
+			for Bond { amount, .. } in &top_nominators {
+				total_counted += *amount;
+				total_backing += *amount;
+			}
+			for Bond { amount, .. } in &bottom_nominators {
+				total_backing += *amount;
+			}
+			// update candidate pool with new total counted if it changed
+			if state.total_counted != total_counted && state.is_active() {
+				reads += 1u64;
+				writes += 1u64;
+				<Pallet<T>>::update_active(account.clone(), total_counted);
```

Comment:
`update_active` does three storage writes as far as I can tell, I dont see we are counting those anywhere, is there something I am missing?

Reply:
+ 1 read